### PR TITLE
fix: async running BQ operations in loop

### PR DESCRIPTION
**Pull request by levibostian**
Closes: https://github.com/customerio/customerio-ios/issues/248

2.0.2 could encounter an infinite loop and eventually stackoverflow while executing the background queue. This was because the BQ while loop would has an infinite loop running while not waiting for the *async HTTP request* to finish. Running async code in a while loop requires some extra logic that was previously not present. 

Automated unit test added that verifies the BQ while loop runs async tasks in sequential order without causing an infinite loop. 

Still todo: 
- [X] QA smoke tests. RH branch `levi/2.0.2-stackoverflow` created to perform testing for this branch. Use builds from that branch.

### CODE REVIEW

File: Sources/Common/Background Queue/QueueRunRequest.swift [hunk 1]
```diff
@@ -38,106 +41,124 @@ public class CioQueueRunRequest: QueueRunRequest {
...
-        else {
-            // we hit the end of the current inventory. Done!
-            return doneRunning()
-        }
+    internal func runTasks() {
+        // Variables that power the logic of the queue run loop
+        var lastRanTask: QueueTaskMetadata?
+        var lastFailedTask: QueueTaskMetadata?
+        var continueRunnning = true
+        let whileLoopWait = DispatchGroup() // make async operations perform synchronously in while loop
```
Comment (Shahroz16) [inline Sources/Common/Background Queue/QueueRunRequest.swift:57]: is there any specific reason we opted for `DispatchGroup` rather `Semaphores` I believe they do serve somewhat a similar purpose?

**Review by Shahroz16 — APPROVED** [review]
Added one comment, the rest looks good.

File: Sources/Common/Background Queue/QueueRunRequest.swift (hunk 1 above)
Comment (levibostian) [reply Sources/Common/Background Queue/QueueRunRequest.swift:57]: Using a semaphore didn't come to mind. In fact, I have never used a semaphore in iOS before. 

`DispatchGroup` is popular in iOS/mac development and will more then likely see the class used. I am not sure about the history of `DispatchGroup`, but I would guess that it's an abstraction that is provided by Apple to provide a more convenient API during development. 

The same could be said about using `DispatchQueue` for creating and executing code on a thread. Or, using `Core Data` for writing to a database. All are abstraction APIs provided by Apple that are popular to use in the development community. However, there are usually alternatives to these abstractions. 

Long story short, `DispatchGroup` is popular, iOS devs are familiar with it, and it's simple to use.
